third_party/amd/lib/TritonAMDGPUTransforms/PointerCanonicalizer.cpp

```diff
@@ -0,0 +1,506 @@
+Value getScalarOffset(IRRewriter &rewriter, Location loc, Value offset) {
+  if (!isa<RankedTensorType>(offset.getType()))
+    return offset;
+
+  if (auto splatOp = dyn_cast<triton::SplatOp>(offset.getDefiningOp()))
+    return splatOp.getSrc();
+
+  if (auto constOp = dyn_cast<arith::ConstantOp>(offset.getDefiningOp())) {
+    auto denseAttr = dyn_cast<DenseIntElementsAttr>(constOp.getValueAttr());
+    if (denseAttr && denseAttr.isSplat())
+      return rewriter.create<arith::ConstantOp>(
+          loc, denseAttr.getSplatValue<IntegerAttr>());
+  }
+
+  return Value();
+}
+} // namespace
+
+// Create a tensor pointer from a fat pointer `fatPtr`. The tensor pointer is
+// obtained by splatting the scalar pointer using the `fatPtr.offset` shape.
+Value PointerCanonicalizer::createTensorPointer(FatPtr fatPtr, Location loc) {
+  Value basePtr = fatPtr.basePtr;
+  Value offset = fatPtr.offset;
+  // Get the offset shape
+  auto offsetType = dyn_cast<RankedTensorType>(offset.getType());
+  ArrayRef<int64_t> offsetShape = offsetType.getShape();
+  // Splat the scalar pointer
+  auto tensorPtrType = RankedTensorType::get(offsetShape, basePtr.getType(),
+                                             offsetType.getEncoding());
+  Value tensorPtr =
+      rewriter.create<triton::SplatOp>(loc, tensorPtrType, basePtr);
+  return tensorPtr;
+}
+
+// Rewrite a memory operation
+LogicalResult PointerCanonicalizer::materializeFatPointer(Operation *op,
+                                                          Location loc,
+                                                          Value ptr) {
+  auto fatPtr = pointers[ptr];
+  Value basePtr = fatPtr.basePtr;
+  Value offset = fatPtr.offset;
+  if (fatPtr.canNarrow)
+    offset = narrow64bitOffsetTo32bits(rewriter, loc, offset);
+
+  Value newPtr = basePtr;
+  if (isa<RankedTensorType>(ptr.getType())) {
+    // Splat the base pointer
+    Value tensorPtr = createTensorPointer(fatPtr, loc);
+    // This is creating `tt.addptr(%tensorBasePtr, %fatPtr.offset)
+    newPtr = rewriter.create<triton::AddPtrOp>(loc, tensorPtr.getType(),
+                                               tensorPtr, offset);
+  }
+
+  // Map and replace the load
+  IRMapping mapper;
+  mapper.map(ptr, newPtr);
+  Operation *newOp = rewriter.clone(*op, mapper);
+  rewriter.replaceAllOpUsesWith(op, newOp);
+  return success();
+}
+
+// Start from an argument of a function and propagate its
+// fat pointers
+LogicalResult PointerCanonicalizer::rewritePointer(Value argPtr) {
+  // List of IR to delete
+  SetVector<Operation *> opToDelete;
+  // This structure is used to point to the right operation during the traversal
+  DenseMap<Operation *, Operation *> rewriteOpMap;
+  // Queue of operations to visit
+  SmallVector<OpOperand *> queue;
+
+  // Start the visit
+  for (OpOperand &use : argPtr.getUses())
+    queue.push_back(&use);
+
+  while (!queue.empty()) {
+    OpOperand *curOperand = queue.back();
+    queue.pop_back();
```

Comment:
Since this is depth first, could also be implemented recursively.. but iterative is fine also.